numpyro/distributions/distribution.py

```diff
@@ -106,6 +106,24 @@ class Distribution(object):
     reparametrized_params = []
     _validate_args = False
 
+    # register Distribution as a pytree
+    # ref: https://github.com/google/jax/issues/2916
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+        tree_util.register_pytree_node(cls,
+                                       cls.tree_flatten,
+                                       cls.tree_unflatten)
+
+    def tree_flatten(self):
+        # XXX: this won't work with MultivariateNormal, which has different interpretations
+        # depending on which one specified among `scale_tril`, `covariance_matrix`, `precision_matrix`
+        # TODO: arrange keys in arg_constraints to be the same order as constructor
+        return tuple(getattr(self, param) for param in self.arg_constraints.keys()), None
```

Comment:
Good point! I think it should work but we need to overwrite the methods in such subclasses. Auxiliary information such as `batch_shape`, `reinterpreted_batch_ndims` can be put into the second output (instead of `None` here) and it will be recovered correctly by the argument `aux_data` in `tree_unflatten`. `mask` can be scanned as other `arg_constraints`. TransformedDistribution is trickier, I think we need to register pytree for them too. We can add try/except here (e.g. to check if `self.arg_constraints` is non-empty, otherwise raise an error to ask users to submit a feature request. :)

Reply:
It might be better to provide `Distribution.flatten` or `Distribution.unflatten` methods if we need to do this, instead of baking in assumptions in `arg_constraints`, though let us hold off on that until we finalize the approach. The good thing about this change is that it is self-contained, and can be easily reverted later.